lib/internal/streams/async_iterator.js

```diff
@@ -0,0 +1,141 @@
+const kHandlePromise = Symbol('handlePromise');
+const kStream = Symbol('stream');
+
+const AsyncIteratorRecord = class AsyncIteratorRecord {
+  constructor(value, done) {
+    this.done = done;
+    this.value = value;
+  }
+}
+
+function readAndResolve(iter) {
+  const resolve = iter[kLastResolve];
+  if (resolve !== null) {
+    const data = iter[kStream].read();
+    // we defer if data is null
+    // we can be expecting either 'end' or
+    // 'error'
+    if (data !== null) {
+      resolve(new AsyncIteratorRecord(data, false));
+      iter[kLastResolve] = null;
+      iter[kLastReject] = null;
+    }
+  }
+}
+
+function onReadable(iter) {
+  // we wait for the next tick, because it might
+  // emit an error with process.nextTick
+  process.nextTick(readAndResolve, iter);
+}
+
+function onEnd(iter) {
+  const resolve = iter[kLastResolve];
+  if (resolve !== null) {
+    resolve(new AsyncIteratorRecord(null, true));
+    iter[kLastResolve] = null;
+    iter[kLastReject] = null;
+  }
+  iter[kEnded] = true;
+}
+
+function onError(iter, err) {
+  const reject = iter[kLastReject];
+  // reject if we are waiting for data in the Promise
+  // returned by next() and store the error
+  if (reject !== null) {
+    reject(err);
+    iter[kLastResolve] = null;
+    iter[kLastReject] = null;
+  }
+  iter.error = err;
+}
+
+const ReadableAsyncIterator = class ReadableAsyncIterator {
+  constructor(stream) {
+    this[kStream] = stream;
+    this[kLastResolve] = null;
+    this[kLastReject] = null;
+    this[kError] = null;
+    this[kEnded] = false;
+    this[kLastPromise] = null;
+
+    stream.on('readable', onReadable.bind(null, this));
+    stream.on('end', onEnd.bind(null, this));
+    stream.on('error', onError.bind(null, this));
+
+    // the function passed to new Promise
+    // is cached so we avoid allocating a new
+    // closure at every run
+    this[kHandlePromise] = (resolve, reject, promise, err) => {
+      const data = this[kStream].read();
+      if (data) {
+        resolve(new AsyncIteratorRecord(data, false));
+      } else {
+        this[kLastResolve] = resolve;
+        this[kLastReject] = reject;
+      }
+    };
+  }
+
+  get stream() {
+    return this[kStream];
+  }
+
+  next() {
+    // if we have detected an error in the meanwhile
+    // reject straight away
+    const error = this[kError];
+    if (error !== null) {
+      return Promise.reject(error);
+    }
+
+    if (this[kEnded]) {
+      return Promise.resolve(new AsyncIteratorRecord(null, true));
+    }
+
+    // if we have multiple next() calls
+    // we will wait for the previous Promise to finish
+    // this logic is optimized to support for await loops,
+    // where next() is only called once at a time
+    const lastPromise = this[kLastPromise];
+    let promise;
+
+    if (lastPromise) {
+      promise = new Promise((resolve, reject) => {
+        lastPromise.then(() => {
+          this[kHandlePromise](resolve, reject)
+        }, reject)
+      });
```

Comment:
This is almost definitely worse for performance since we're creating extra closures and `lastPromise` is never unset, so in a `for..await` loop this will always `then` even though it doesn't need to.

Reply:
> so in a for..await loop this will always then even though it doesn't need to.

This is not correct. However it was showing that behavior in the previous implementation.

Reply:
> However it was showing that behavior in the previous implementation.

Yeah, you're right. I think I had accounted for unsetting `lastPromise` if several `read` calls were scheduled but not in `for..await`.

I think you can check if `lastResolve` is set before setting `this[kLastPromise]`.

Reply:
See prior discussion @apapirovski - we optimize for the fast (for...await) case while supporting the slower (non-for-await) case where multiple `next`s are called.

Reply:
@benjamingr But we don't? `this[kLastPromise]` is always set... so we always call `then`. As far as I can tell, the code can be rewritten as such:

```js
if (lastResolve !== null) {
  promise = new Promise(wrapForNext(lastPromise, this));
} else {
  promise = new Promise(this[kHandlePromise]);
}
if (lastResolve !== null)
  this[kLastPromise] = promise;
```

That will account for not doing an unnecessary `then` for an already resolved `lastPromise`.

Reply:
Should be handled now, please check.